### ISSUE TITLE
Large memory consumption

### ISSUE THREAD

**Issue by npmhung** [michaelsdr/momentumnet#16]
I tested the memory consumption of resnet 101 and the corresponding momentum version using the scripts below. However, difference in memory consumption between these 2 models is only a few hundreds MBs.

Pytorch version: 1.9
CUDA toolkit: 11.1
GPU: GTX 1080 Ti
Python 3.7
Ubuntu 18.04

```import torch
from momentumnet import transform_to_momentumnet
from torchvision.models import resnet101

resnet = resnet101(pretrained=True).to(0)
mresnet101 = transform_to_momentumnet(resnet, gamma=0.99998, use_backprop=False)
#mresnet101 = resnet101(pretrained=True).to(0)
optim = torch.optim.Adam(mresnet101.parameters(), lr=1e-4)

for i in range(100):
    x = torch.randn(70, 3, 224, 224).cuda()
    y = mresnet101(x)
    y = (y**2).mean()
    optim.zero_grad()
    y.backward()
    optim.step()

```

**Comment by michaelsdr** [comment]
Hello, I guess that is because you need to restart the kernel between two memory profiles 
For instance you can try
`
import torch
from momentumnet import transform_to_momentumnet
from torchvision.models import resnet152


device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

X = torch.randn(70, 3, 224, 224).cuda()

def train(net):
    Loss = (net(X) ** 2).mean()
    Loss.backward()

resnet = resnet152(pretrained=False).to(0)
mom_net = transform_to_momentumnet(resnet, gamma=0.99, use_backprop=False)

train(resnet)
used_mem = torch.cuda.max_memory_allocated(device=device)
print('For classical ResNet: ', used_mem)
`

and then **after restarting the kernel**

`
import torch
from momentumnet import transform_to_momentumnet
from torchvision.models import resnet152


device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

X = torch.randn(70, 3, 224, 224).cuda()

def train(net):
    Loss = (net(X) ** 2).mean()
    Loss.backward()

resnet = resnet152(pretrained=False).to(0)
mom_net = transform_to_momentumnet(resnet, gamma=0.99, use_backprop=False)


torch.cuda.empty_cache()
train(mom_net)
used_mem = torch.cuda.max_memory_allocated(device=device)
print('For the MomentumNet: ', used_mem)
`

**Comment by michaelsdr** [comment]
```
import torch
from momentumnet import transform_to_momentumnet
from torchvision.models import resnet152


device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

X = torch.randn(70, 3, 224, 224).cuda()

def train(net):
    Loss = (net(X) ** 2).mean()
    Loss.backward()

resnet = resnet152(pretrained=False).to(0)
mom_net = transform_to_momentumnet(resnet, gamma=0.99, use_backprop=False)


torch.cuda.empty_cache()
train(mom_net)
used_mem = torch.cuda.max_memory_allocated(device=device)
print('For the MomentumNet: ', used_mem)

```

**Comment by michaelsdr** [comment]
Hello, I guess that's because you need to restart the kernel between two trials

For me doing the following works well:

```
import torch
from momentumnet import transform_to_momentumnet
from torchvision.models import resnet152


device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

X = torch.randn(70, 3, 224, 224).cuda()

def train(net):
    Loss = (net(X) ** 2).mean()
    Loss.backward()

resnet = resnet152(pretrained=False).to(0)
mom_net = transform_to_momentumnet(resnet, gamma=0.99, use_backprop=False)

train(resnet)
used_mem = torch.cuda.max_memory_allocated(device=device)
print('For classical ResNet: ', used_mem)
```

**And then after restarting the kernel** 

```
import torch
from momentumnet import transform_to_momentumnet
from torchvision.models import resnet152


device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

X = torch.randn(70, 3, 224, 224).cuda()

def train(net):
    Loss = (net(X) ** 2).mean()
    Loss.backward()

resnet = resnet152(pretrained=False).to(0)
mom_net = transform_to_momentumnet(resnet, gamma=0.99, use_backprop=False)


torch.cuda.empty_cache()
train(mom_net)
used_mem = torch.cuda.max_memory_allocated(device=device)
print('For the MomentumNet: ', used_mem)
```
